Ai improve large map performance

ron-murhammer opened the pull request:
## Overview
- Address really slow AI performance on new alpha map World of War Heroes: https://forums.triplea-game.org/topic/1031/world-of-war-heroes-official-thread/22

## Functional Changes
- Decrease random start delegate pause time from 250ms to 10ms
- Add new get neighbors ignoring end methods to reduce trying to find routes for sea territories that either don't have a valid route or have a very long route
- Update territory value utils to avoid ever checking every territory on the map and always pass in what territories to check. This essentially makes AI calculations more 'regional' instead of 'global' so that the AI only calculates things where it can move/place units. An example is avoid having Japan calculate a bunch of territory values for things in the European Theater when it doesn't have any units that can reach that.

## Manual Testing Performed
- Did functional and performance testing on World of War Heroes alpha, WaW, Dom NML, and Another World. Most large maps with Fast AI are about 2x as fast now but World of War or other sparse large maps are 5x-10x faster.

ron-murhammer commented on game-core/src/main/java/games/strategy/engine/data/GameMap.java:
```diff
@@ -190,7 +192,7 @@ public Territory getTerritory(final String s) {
   private Set<Territory> getNeighbors(final Set<Territory> frontier, final Set<Territory> searched, final int distance,
       @Nullable final Predicate<Territory> cond) {
-    if (distance == 0) {
+    if (distance == 0 || frontier.isEmpty()) {
```
Add empty check here as once the frontier is empty, you won't ever find any more territories to search.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java:
```diff
@@ -49,13 +49,6 @@ public static double findTerritoryAttackValue(final PlayerId player, final Terri
     return value;
   }
 
-  public static Map<Territory, Double> findTerritoryValues(final PlayerId player,
```
This is one of the main keys of this PR was to remove this method and refactor areas that used it as its very slow on large maps.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java:
```diff
@@ -92,12 +84,12 @@ public static double findTerritoryAttackValue(final PlayerId player, final Terri
    */
   public static Map<Territory, Double> findSeaTerritoryValues(final PlayerId player,
-      final List<Territory> territoriesThatCantBeHeld) {
+      final List<Territory> territoriesThatCantBeHeld, final List<Territory> territoriesToCheck) {
```
Always pass in a list of territories to check to avoid checking every territory on the map.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProTerritoryValueUtils.java:
```diff
@@ -302,7 +294,8 @@ private static double findWaterValue(final Territory t, final PlayerId player, f
     double nearbyLandValue = 0;
-    final Set<Territory> nearbyTerritories = data.getMap().getNeighbors(t, 3);
+    final Set<Territory> nearbyTerritories =
+        data.getMap().getNeighborsIgnoreEnd(t, 3, ProMatches.territoryCanMoveSeaUnits(player, data, true));
```
This change is important because often you have cases where 2 sea territories where within 3 hops but they either aren't connected by sea territories at all or the route over sea territories is very long.

ron-murhammer commented:
This pull request **introduces 1 alert** when merging afd24d2d0cdf718ccebfab1b1edf7131a70a918d into f7745891f46673a11dd4493b09cef22dabef2b93 - [view on LGTM.com](https://lgtm.com/projects/g/triplea-game/triplea/rev/pr-074dd36921040d4a7becbee6a7d95dd24573ded3)

**new alerts:**

* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ai/pro/ProCombatMoveAi.java:
```diff
@@ -375,14 +367,25 @@ private void determineTerritoriesToAttack(final List<ProTerritory> prioritizedTe
     final ProOtherMoveOptions enemyAttackOptions = territoryManager.getEnemyAttackOptions();
     final Map<Territory, ProTerritory> attackMap = territoryManager.getAttackOptions().getTerritoryMap();
 
     // Determine which territories to try and hold
```
I'm bit curious if it's easy to explain why the previous method was slow. I assume this added code is a re-write of the slow code.

Is this current/new code still somewhat slow? It looks like a candidate for parallelism. FWIW, this is equivalent down until `territoryValueMap`, and iterates in parallel

```
    // Determine which territories to try and hold
    final Set<Territory> territoriesToCheck = prioritizedTerritories.stream()
        .map(ProTerritory::getTerritory)
        .collect(Collectors.toSet());
    // add all nonAirAttackers from prioritized territories
    territoriesToCheck.addAll(
        prioritizedTerritories
            .parallelStream()
            .map(patd -> CollectionUtils.getMatches(patd.getMaxUnits(), Matches.unitIsNotAir()))
            .flatMap(Collection::stream)
            .map(nonAirAttacker -> ProData.unitTerritoryMap.get(nonAirAttacker))
            .collect(Collectors.toSet()));
```

ron-murhammer replied:
So this portion of the code is very fast and was added to determine which territories need to be passed to TerritoryValueUtils which is the slow method.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/GameMap.java:
```diff
@@ -207,6 +209,46 @@ public Territory getTerritory(final String s) {
+  public Set<Territory> getNeighborsIgnoreEnd(final Territory territory, final int distance,
+      final Predicate<Territory> cond) {
+    if (distance < 0) {
+      throw new IllegalArgumentException("Distance must be positive not: " + distance);
```
really nit-picking, 'zero' is not a positive number and distnace can be zero.

Suggestion, as a one-liner with guava:
```
Preconditions.checkArgument(distance < 0, "Distance must be non-negative: " + distance);
```

ron-murhammer replied:
Updated.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/GameMap.java:
```diff
@@ -207,6 +209,46 @@ public Territory getTerritory(final String s) {
   }
 
+  /**
+   * Returns all neighbors within a certain distance of the starting territory that territories
```
The first sentence could use a re-read. Should it read something like the following (?): 

>  certain distance of the starting territory [where all territories] between the 2 match a condition.

ron-murhammer replied:
Agree. Updated.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/GameMap.java:
```diff
@@ -207,6 +209,46 @@ public Territory getTerritory(final String s) {
+
+  private Set<Territory> getNeighborsIgnoreEnd(final Set<Territory> frontier, final Set<Territory> searched,
+      final int distance, @Nullable final Predicate<Territory> cond) {
+    if (distance == 0) {
```
Very minor: A comment that this condition terminates recursion could be a good heads-up for the future maintenance reviewer. I was wondering a bit when this could be zero to eventually see it could be after recursive iteration.

ron-murhammer replied:
Updated.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/data/GameMap.java:
```diff
@@ -207,6 +209,46 @@ public Territory getTerritory(final String s) {
+      return neighbors;
+    }
+    final Set<Territory> start = getNeighbors(territory, cond);
+    for (int i = 2; i <= distance; i++) {
```
Minor: I wonder a bit why the value '2'. Extracting it to a constant and adding a comment on it could be a good way to document the reasoning why the number '2' is here. I would guess it is so we do not consider too many territories, but not sure if maybe the reason is because we can assume movement is limited.

ron-murhammer replied:
You start with 2 since there are specific conditions already for <0, 0, and 1. So if distance is 2 or more than you start doing the recursion calls for each distance. I don't think a constant would really make sense here.

DanVanAtta replied:
Okay, I wonder a bit why not go up to 3 or higher, but I'll take this at face-value. :+1:

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ai/pro/ProNonCombatMoveAi.java:
```diff
@@ -526,6 +525,12 @@ private void determineIfMoveTerritoriesCanBeHeld() {
     seaFactories = CollectionUtils.getMatches(seaFactories,
         ProMatches.territoryHasInfraFactoryAndIsOwnedLandAdjacentToSea(player, data));
+    final Set<Territory> territoriesToCheck = new HashSet<>(seaFactories);
+    for (final Territory t : seaFactories) {
```
This looks to be another parallel stream opportunity. I'm not sure how much it helps performance here, but FWIW here is the parallel stream equivalent:
```
    territoriesToCheck.addAll(
        seaFactories.parallelStream()
            .map(t -> data.getMap().getNeighbors(
                t, ProMatches.territoryCanMoveSeaUnits(player, data, true)))
            .flatMap(Collection::stream)
            .collect(Collectors.toSet()));
```

ron-murhammer replied:
Doesn't really make a difference here as this method is super fast and just finding which territories need checked for TerritoryValueUtils.

DanVanAtta commented on game-core/src/main/java/games/strategy/triplea/ai/pro/ProNonCombatMoveAi.java:
```diff
@@ -878,6 +882,18 @@ private void moveUnitsToDefendTerritories(final List<ProTerritory> prioritizedTe
       boolean areSuccessful = true;
       boolean containsCapital = false;
+      final Set<Territory> territoriesToCheck = new HashSet<>();
+      for (final ProTerritory patd : territoriesToTryToDefend) {
```
Noting that we have a matches operation, maybe that could benefit from parallelism? Not sure, FWIW a quick parallel version:
```
      territoriesToTryToDefend.parallelStream()
          .forEach(patd -> {
            final Territory t = patd.getTerritory();
            territoriesToCheck.add(t);
            final List<Unit> nonAirDefenders =
                CollectionUtils.getMatches(moveMap.get(t).getTempUnits(), Matches.unitIsNotAir());
            for (final Unit u : nonAirDefenders) {
              territoriesToCheck.add(unitTerritoryMap.get(u));
            }
          });
```

ron-murhammer replied:
Same as above.

DanVanAtta commented:
Looks good @ron-murhammer . Left a few comments/suggestions, let me know what you think and when done iterating.

ron-murhammer commented on game-core/src/main/java/games/strategy/triplea/delegate/RandomStartDelegate.java:
```diff
@@ -94,7 +94,7 @@ private void setupBoard() {
       if (currentPickingPlayer == null || !playersCanPick.contains(currentPickingPlayer)) {
         currentPickingPlayer = playersCanPick.get(0);
       }
-      if (!Interruptibles.sleep(250)) {
```
I actually don't know if its necessary which is why I didn't just remove it entirely but left a small sleep in case their is some reason from a painting the UI tiles that needs it. Probably better as a follow up if someone wants to test just removing it entirely and testing a bunch of maps.

DanVanAtta replied:
If you're feeling up for a it, a TODO comment with that question/context could be a good placeholder .

ron-murhammer replied:
Good idea. Added.

ron-murhammer commented:
@DanVanAtta Addressed comments and made some updates. If you want to take a look at where the performance is slow and potentially where a parallel stream could help it would be in the `ProTerritoryValueUtils` like this method: https://github.com/triplea-game/triplea/pull/4764/files#diff-0f55e4b4525775beb799afb6b97cc2b9R265

ron-murhammer commented:
This pull request **introduces 1 alert** when merging d3e34b38476ea5c904402521d2a661c47c16c9e9 into f7745891f46673a11dd4493b09cef22dabef2b93 - [view on LGTM.com](https://lgtm.com/projects/g/triplea-game/triplea/rev/pr-1319c57d5a0ec430d63b095148bf79c60c4cece2)

**new alerts:**

* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

ron-murhammer commented:
This pull request **introduces 1 alert** when merging a750a67320f9732b63e7c0c11cc394d03669f0d5 into f7745891f46673a11dd4493b09cef22dabef2b93 - [view on LGTM.com](https://lgtm.com/projects/g/triplea-game/triplea/rev/pr-46494f75a3991c79fe1ba7f5e007cb1409bd063e)

**new alerts:**

* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

DanVanAtta commented:
@ron-murhammer any idea if the LGTM alert is valid or should be suppressed?

This is looking pretty updated/addressed, I think that may be the last thing.

ron-murhammer commented:
@DanVanAtta That is actually valid and a really good catch. Fixed.
